  # This notebook will be used for Coursera Capstone Project

## Assignment week 3: Explore and cluster the neighborhoods in Toronto.

### We have webscrape the toronto website and obtain a clean dataframe

<p> Web scrapping is a technique to automatically access and extract large amount of information from a website,
to save huge amount of time and effort <p>

<p> Use the Notebook to build the code to scrape the following website 
<a href="https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M">https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M</a> in order to obtain 
        the data that is in the table of postal codes and to transform the data into a pandas dataframe <p>


In [6]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import requests
import urllib.request
import time
import csv
from bs4 import BeautifulSoup

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

### enter the url of the site to scrape

In [20]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
response = requests.get(url)

In [21]:
soup = BeautifulSoup(response.text,"html.parser")

In [73]:
#print(soup.prettify())

### Initialize the lists and use find_all method to find our table tags

In [22]:
postalCodeList = []
boroughList = []
neighborhoodList = []
table = soup.find('table').find_all('tr')
#print(table)
for row in table:
        cells = row.find_all('td')
        if(len(cells) > 0):
            postalCodeList.append(cells[0].text)
            boroughList.append(cells[1].text)
            neighborhoodList.append(cells[2].text.rstrip('\n')) #remove new line character from cell
            

            
    
 

In [23]:
# create a dataframe
toronto_data = [('PostalCode', postalCodeList),
                     ('Borough', boroughList),
                    ('Neighborhood', neighborhoodList)]
toronto_df = pd.DataFrame.from_dict(dict(toronto_data))
toronto_df.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

<b>Remove not assigned rows </b> <p> There are some postal codes which are not belongs to any borough.
We will remove those rows.</p>

In [24]:
toronto_df_dropna = toronto_df[toronto_df.Borough != 'Not assigned'].reset_index(drop=True)
toronto_df_dropna.head()

PostalCode           Borough      Neighborhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M5A  Downtown Toronto       Regent Park
4        M6A        North York  Lawrence Heights

<b> Group neighborhoods by postal and borough </b>
<p>There are some neighborhoods that belongs to the same postal code and borough (Ex: M5A Downtown Toronto has 2 neighborhoods Harbourfront and Regent Park).
We will concat them into same row, seperated by a colon.<p>

In [25]:
toronto_df_grouped = toronto_df_dropna.groupby(['PostalCode','Borough'], as_index=False).agg(lambda x: ','.join(x))
toronto_df_grouped.head()

PostalCode      Borough                          Neighborhood
0        M1B  Scarborough                         Rouge,Malvern
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough       Guildwood,Morningside,West Hill
3        M1G  Scarborough                                Woburn
4        M1H  Scarborough                             Cedarbrae

<b> Not Assigned Neighbourhood </b>
<p>If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.</p>

In [26]:
na_neigh_rows = toronto_df_grouped.Neighborhood == 'Not assigned'
toronto_df_grouped.loc[na_neigh_rows, 'Neighborhood'] = toronto_df_grouped.loc[na_neigh_rows, 'Borough']
toronto_df_grouped[na_neigh_rows]

PostalCode       Borough  Neighborhood
85        M7A  Queen's Park  Queen's Park

## A Cleaned Dataframe

In [27]:
toronto_df_cleaned = toronto_df_grouped
toronto_df_cleaned.shape

(103, 3)

### So we have 103 rows in the clean final dataframe

# Assignment part 2 of Week 3

<p> Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.</p>

In [29]:
conda install -c menpo wget

Solving environment: ...working... done

# All requested packages already installed.


Note: you may need to restart the kernel to use updated packages.


In [8]:

!wget -q -O "Geospatial_coordinates.csv" http://cocl.us/Geospatial_data
print('Coordinates downloaded!')



In [17]:
coors = pd.read_csv('Geospatial_coordinates.csv')

In [18]:
print(coors.shape)
coors.head()

(103, 3)


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [28]:
toronto_df_temp = toronto_df_cleaned.set_index('PostalCode')
coors_temp = coors.set_index('Postal Code')
toronto_df_coors = pd.concat([toronto_df_temp, coors_temp], axis=1, join='inner')

In [29]:
toronto_df_coors.index.name = 'PostalCode'
toronto_df_coors.reset_index(inplace=True)

In [30]:
print(toronto_df_coors.shape)
toronto_df_coors.head()

(103, 5)


PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

## Part 3: Explore and cluster the neighborhoods in Toronto

In [31]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="tl-toronto-neigh")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.653963, -79.387207.


## Using Folium to create a Map of Toronto with Boroughs markers on top

In [33]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, long, post, borough, neigh in zip(toronto_df_coors['Latitude'], toronto_df_coors['Longitude'], toronto_df_coors['PostalCode'], toronto_df_coors['Borough'], toronto_df_coors['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='red',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

<b>Reduce the number of Boroughs</b><p> to explore To reduce the numbers of calls to FourSquare API, we will only explore boroughs that have Toronto in their names.</p>

In [34]:
toronto_boroughs = ['East Toronto', 'Central Toronto', 'Downtown Toronto', 'West Toronto']
toronto_central_df = toronto_df_coors[toronto_df_coors['Borough'].isin(toronto_boroughs)].reset_index(drop=True)
print(toronto_central_df.shape)
toronto_central_df.head()

(38, 5)


PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [36]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

for lat, long, post, borough, neigh in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='red',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

### Using FourSquare API to explore the Boroughs

<p>Get recommend places inside or near each Borough in Toronto Central</p>

<b>FourSquare Credentials</b>

In [38]:
CLIENT_ID = 'NV2NVCE044GXDUDCLOWI0RGOC1BXDMTZM5UAT14F3HWLBBC1' # your Foursquare ID
CLIENT_SECRET = '3JH04HHEVYQOGQBXEW3XEKAYW4RDP5TKVVZEU1KXECEVDFZP' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: NV2NVCE044GXDUDCLOWI0RGOC1BXDMTZM5UAT14F3HWLBBC1
CLIENT_SECRET:3JH04HHEVYQOGQBXEW3XEKAYW4RDP5TKVVZEU1KXECEVDFZP


In [39]:
radius = 500
LIMIT = 100

venues = []

for lat, long, post, borough, neighborhood in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    for venue in results:
        venues.append((
            post, 
            borough,
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [40]:
venues_df = pd.DataFrame(venues)
venues_df.columns = ['PostalCode', 'Borough', 'Neighborhood', 'BoroughLatitude', 'BoroughLongitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']
print(venues_df.shape)
venues_df.head()

(1678, 9)


PostalCode       Borough Neighborhood  BoroughLatitude  BoroughLongitude  \
0        M4E  East Toronto  The Beaches        43.676357        -79.293031   
1        M4E  East Toronto  The Beaches        43.676357        -79.293031   
2        M4E  East Toronto  The Beaches        43.676357        -79.293031   
3        M4E  East Toronto  The Beaches        43.676357        -79.293031   
4        M4E  East Toronto  The Beaches        43.676357        -79.293031   

                            VenueName  VenueLatitude  VenueLongitude  \
0                   Glen Manor Ravine      43.676821      -79.293942   
1  The Big Carrot Natural Food Market      43.678879      -79.297734   
2                 Grover Pub and Grub      43.679181      -79.297215   
3                   Glen Stewart Park      43.675278      -79.294647   
4                 Glen Stewart Ravine      43.676300      -79.294784   

          VenueCategory  
0                 Trail  
1     Health Food Store  
2                   Pub  
3                  Park  
4  Other Great Outdoors

## Check how many venues returned for each postal code

In [41]:
venues_df.groupby(['PostalCode', 'Borough', 'Neighborhood'])['VenueName'].count()

PostalCode  Borough           Neighborhood                                                                                        
M4E         East Toronto      The Beaches                                                                                               6
M4K         East Toronto      The Danforth West,Riverdale                                                                              42
M4L         East Toronto      The Beaches West,India Bazaar                                                                            19
M4M         East Toronto      Studio District                                                                                          37
M4N         Central Toronto   Lawrence Park                                                                                             4
M4P         Central Toronto   Davisville North                                                                                          7
M4R         Central Toronto   North Toron

## Check how many kinds of venues categories is returned

In [42]:
len(venues_df['VenueCategory'].unique())

229

### Analyze venues in each area

In [43]:
# one hot encoding
toronto_central_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add postal, borough and neighborhood column back to dataframe
toronto_central_onehot['PostalCode'] = venues_df['PostalCode'] 
toronto_central_onehot['Borough'] = venues_df['Borough'] 
toronto_central_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move postal, borough and neighborhood column to the first column
fixed_columns = list(toronto_central_onehot.columns[-3:]) + list(toronto_central_onehot.columns[:-3])
toronto_central_onehot = toronto_central_onehot[fixed_columns]

print(toronto_central_onehot.shape)
toronto_central_onehot.head()

(1678, 232)


PostalCode       Borough Neighborhoods  Afghan Restaurant  Airport  \
0        M4E  East Toronto   The Beaches                  0        0   
1        M4E  East Toronto   The Beaches                  0        0   
2        M4E  East Toronto   The Beaches                  0        0   
3        M4E  East Toronto   The Beaches                  0        0   
4        M4E  East Toronto   The Beaches                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       0     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0                   0       0           0              0          0         0   
1                   0       0           0              0          0         0   
2                   0       0           0              0          0         0   
3                   0       0           0              0          0         0   
4                   0       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0            

### Get the frequency of occurance of each category in an area

In [44]:
toronto_central_venues_freq = toronto_central_onehot.groupby(['PostalCode', 'Borough', 'Neighborhoods']).mean().reset_index()
print(toronto_central_venues_freq.shape)
toronto_central_venues_freq.head()

(38, 232)


PostalCode          Borough                  Neighborhoods  \
0        M4E     East Toronto                    The Beaches   
1        M4K     East Toronto    The Danforth West,Riverdale   
2        M4L     East Toronto  The Beaches West,India Bazaar   
3        M4M     East Toronto                Studio District   
4        M4N  Central Toronto                  Lawrence Park   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.023810   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.054054   
4             0.0              0.0               0.0             0.000000   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0           0.0       0.0          0.0                  0.0               0.0   
1           0.0       0.0          0.0                  0.0               0.0   
2           0.0       0.0          0.0                  0.0               0.0   
3           0.0       0.0          0.0                  0.0               0.0   
4           0.0       0.0          0.0                  0.0               0.0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                 0.0            0.0        0.0         0.0         0.0   
1                 0.0            0.0        0.0         0.0         0.0   
2                 0.0            0.0        0.0         0.0         0.0   
3                 0.0            0.0        0.0         0.0         0.0   
4                 0.0            0.0        0.0         0.0         0.0   

     Bakery      Bank       Bar  Baseball Stadium  Basketball Stadium  Beach  \
0  0.000000  0.000000  0.000000               0.0                 0.0    0.0   
1  0.023810  0.000000  0.000000               0.0                 0.0    0.0   
2  0.000000  0.000000  0.000000               0.0                 0.0    0.0   
3  0.054054  0.027027  0.027027               0.0                 0.0    0.0   
4  0.000000  0.000000  0.000000               0.0                 0.0    0.0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0              0.0       0.0         0.0                 0.0     0.0   
1              0.0       0.0         0.0                 0.0     0.0   
2              0.0       0.0         0.0                 0.0     0.0   
3              0.0       0.0         0.0                 0.0     0.0   
4              0.0       0.0         0.0                 0.0     0.0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0    0.000000            0.0   0.000000       0.0                   0.0   
1    0.000000            0.0   0.023810       0.0                   0.0   
2    0.052632            0.0   0.000000       0.0                   0.0   
3    0.000000            0.0   0.027027       0.0                   0.0   
4    0.000000            0.0   0.000000       0.0                   0.0   

   Breakfast Spot   Brewery  Bubble Tea Shop  Building  Burger Joint  \
0             0.0  0.000000          0.00000       0.0      0.000000   
1             0.0  0.023810          0.02381       0.0      0.000000   
2             0.0  0.052632          0.00000       0.0      0.052632   
3             0.0  0.027027          0.00000       0.0      0.000000   
4             0.0  0.000000          0.00000       0.0      0.000000   

   Burrito Plac

### Get 10 most occurance venue types in each area

In [45]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
areaColumns = ['PostalCode', 'Borough', 'Neighborhoods']
freqColumns = []
for ind in np.arange(num_top_venues):
    try:
        freqColumns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        freqColumns.append('{}th Most Common Venue'.format(ind+1))
columns = areaColumns+freqColumns
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['PostalCode'] = toronto_central_venues_freq['PostalCode']
neighborhoods_venues_sorted['Borough'] = toronto_central_venues_freq['Borough']
neighborhoods_venues_sorted['Neighborhoods'] = toronto_central_venues_freq['Neighborhoods']

for ind in np.arange(toronto_central_venues_freq.shape[0]):
    row_categories = toronto_central_venues_freq.iloc[ind, :].iloc[3:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    neighborhoods_venues_sorted.iloc[ind, 3:] = row_categories_sorted.index.values[0:num_top_venues]

neighborhoods_venues_sorted.sort_values(freqColumns, inplace=True)
neighborhoods_venues_sorted

PostalCode           Borough  \
27        M5V  Downtown Toronto   
32        M6J      West Toronto   
4         M4N   Central Toronto   
26        M5T  Downtown Toronto   
3         M4M      East Toronto   
25        M5S  Downtown Toronto   
14        M5B  Downtown Toronto   
6         M4R   Central Toronto   
33        M6K      West Toronto   
18        M5H  Downtown Toronto   
21        M5L  Downtown Toronto   
20        M5K  Downtown Toronto   
15        M5C  Downtown Toronto   
36        M6S      West Toronto   
29        M5X  Downtown Toronto   
28        M5W  Downtown Toronto   
19        M5J  Downtown Toronto   
11        M4X  Downtown Toronto   
17        M5G  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
16        M5E  Downtown Toronto   
7         M4S   Central Toronto   
22        M5N   Central Toronto   
35        M6R      West Toronto   
1         M4K      East Toronto   
30        M6G  Downtown Toronto   
5         M4P   Central Toronto   
37        M7Y      East Toronto   
34        M6P      West Toronto   
2         M4L      East Toronto   
0         M4E      East Toronto   
10        M4W  Downtown Toronto   
23        M5P   Central Toronto   
31        M6H      West Toronto   
8         M4T   Central Toronto   
9         M4V   Central Toronto   
24        M5R   Central Toronto   

                                        Neighborhoods 1st Most Common Venue  \
27  CN Tower,Bathurst Quay,Island airport,Harbourf...        Airport Lounge   
32                            Little Portugal,Trinity                   Bar   
4                                       Lawrence Park              Bus Line   
26            Chinatown,Grange Park,Kensington Market                  Café   
3                                     Studio District                  Café   
25                      Harbord,University of Toronto                  Café   
14                            Ryerson,Garden District        Clothing Store   
6                                  North Toronto West        Clothing Store   
33         Brockton,Exhibition Place,Parkdale Village           Coffee Shop   
18                             Adelaide,King,Richmond           Coffee Shop   
21                      Commerce Court,Victoria Hotel           Coffee Shop   
20            Design Exchange,Toronto Dominion Centre           Coffee Shop   
15                                     St. James Town           Coffee Shop   
36                                  Runnymede,Swansea           Coffee Shop   
29              First Canadian Place,Underground city           Coffee Shop   
28                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
19    Harbourfront East,Toronto Islands,Union Station           Coffee Shop   
11                         Cabbagetown,St. James Town           Coffee Shop   
17                                 Central Bay Street           Coffee Shop   
12                               Church and Wellesley           Coffee Shop   
13                           Harbourfront,Regent Park           Coffee Shop   
16                                        Berczy Park           Coffee Shop   
7                                          Davisville          Dessert Shop   
22                                           Roselawn                Garden   
35                              Parkdale,Roncesvalles             Gift Shop   
1                         The Danforth West,Riverdale      Greek Restaurant   
30                                           Christie         Grocery Store   
5                                    Davisville North                   Gym   
37  Business Reply Mail Processing Centre 969 Eastern  Gym / Fitness Center   
34                       High Park,The Junction South    Mexican Restaurant   
2                       The Beaches West,India Bazaar                  Park   
0                                         The Beaches                  Park   
10                            

### Clustering areas
<p>Use KMeans algorigthm, try to cluster the toronto central areas into 3 clusters</p>

In [46]:
kclusters = 3

toronto_central_venues_freq_clustering = toronto_central_venues_freq.drop(['PostalCode', 'Borough', 'Neighborhoods'], 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_central_venues_freq_clustering)

toronto_central_clustered_df = toronto_central_df
toronto_central_clustered_df['Cluster'] = kmeans.labels_

toronto_central_clustered_df = toronto_central_clustered_df.join(neighborhoods_venues_sorted.drop(['Borough', 'Neighborhoods'], 1).set_index('PostalCode'), on='PostalCode')
toronto_central_clustered_df.sort_values(['Cluster'] + freqColumns, inplace=True)
toronto_central_clustered_df

PostalCode           Borough  \
27        M5V  Downtown Toronto   
32        M6J      West Toronto   
4         M4N   Central Toronto   
26        M5T  Downtown Toronto   
3         M4M      East Toronto   
25        M5S  Downtown Toronto   
14        M5B  Downtown Toronto   
6         M4R   Central Toronto   
33        M6K      West Toronto   
18        M5H  Downtown Toronto   
21        M5L  Downtown Toronto   
20        M5K  Downtown Toronto   
15        M5C  Downtown Toronto   
36        M6S      West Toronto   
29        M5X  Downtown Toronto   
28        M5W  Downtown Toronto   
19        M5J  Downtown Toronto   
11        M4X  Downtown Toronto   
17        M5G  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
16        M5E  Downtown Toronto   
7         M4S   Central Toronto   
35        M6R      West Toronto   
1         M4K      East Toronto   
30        M6G  Downtown Toronto   
5         M4P   Central Toronto   
37        M7Y      East Toronto   
34        M6P      West Toronto   
2         M4L      East Toronto   
0         M4E      East Toronto   
31        M6H      West Toronto   
9         M4V   Central Toronto   
24        M5R   Central Toronto   
10        M4W  Downtown Toronto   
23        M5P   Central Toronto   
8         M4T   Central Toronto   
22        M5N   Central Toronto   

                                         Neighborhood   Latitude  Longitude  \
27  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420   
32                            Little Portugal,Trinity  43.647927 -79.419750   
4                                       Lawrence Park  43.728020 -79.388790   
26            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049   
3                                     Studio District  43.659526 -79.340923   
25                      Harbord,University of Toronto  43.662696 -79.400049   
14                            Ryerson,Garden District  43.657162 -79.378937   
6                                  North Toronto West  43.715383 -79.405678   
33         Brockton,Exhibition Place,Parkdale Village  43.636847 -79.428191   
18                             Adelaide,King,Richmond  43.650571 -79.384568   
21                      Commerce Court,Victoria Hotel  43.648198 -79.379817   
20            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576   
15                                     St. James Town  43.651494 -79.375418   
36                                  Runnymede,Swansea  43.651571 -79.484450   
29              First Canadian Place,Underground city  43.648429 -79.382280   
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846   
19    Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752   
11                         Cabbagetown,St. James Town  43.667967 -79.367675   
17                                 Central Bay Street  43.657952 -79.387383   
12                               Church and Wellesley  43.665860 -79.383160   
13                           Harbourfront,Regent Park  43.654260 -79.360636   
16                                        Berczy Park  43.644771 -79.373306   
7                                          Davisville  43.704324 -79.388790   
35                              Parkdale,Roncesvalles  43.648960 -79.456325   
1                         The Danforth West,Riverdale  43.679557 -79.352188   
30                                           Christie  43.669542 -79.422564   
5                                    Davisville North  43.712751 -79.390197   
37  Business Reply Mail Processing Centre 969 Eastern  43.662744 -79.321558   
34                       High Park,The Junction South  43.661608 -79.464763   
2                       The Beaches West,India Bazaar  43.668999 -79.315572   
0                                         The Beaches  43.676357 -79.293031   
31                        Dovercourt Village,Dufferin  43.669005 -79.442259   
9   Deer Park,Forest Hill SE,R

### Show those clusters onto a map

In [47]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, post, bor, poi, cluster in zip(toronto_central_clustered_df['Latitude'], toronto_central_clustered_df['Longitude'], toronto_central_clustered_df['PostalCode'], toronto_central_clustered_df['Borough'], toronto_central_clustered_df['Neighborhood'], toronto_central_clustered_df['Cluster']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, post, poi, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Cluster 0

In [49]:
toronto_central_clustered_df.loc[toronto_central_clustered_df['Cluster'] == 0, toronto_central_clustered_df.columns[[1] + list(range(5, toronto_central_clustered_df.shape[1]))]]

Borough  Cluster 1st Most Common Venue 2nd Most Common Venue  \
27  Downtown Toronto        0        Airport Lounge       Airport Service   
32      West Toronto        0                   Bar           Coffee Shop   
4    Central Toronto        0              Bus Line                  Park   
26  Downtown Toronto        0                  Café                   Bar   
3       East Toronto        0                  Café           Coffee Shop   
25  Downtown Toronto        0                  Café            Restaurant   
14  Downtown Toronto        0        Clothing Store           Coffee Shop   
6    Central Toronto        0        Clothing Store   Sporting Goods Shop   
33      West Toronto        0           Coffee Shop        Breakfast Spot   
18  Downtown Toronto        0           Coffee Shop                  Café   
21  Downtown Toronto        0           Coffee Shop                  Café   
20  Downtown Toronto        0           Coffee Shop                  Café   
15  Downtown Toronto        0           Coffee Shop                  Café   
36      West Toronto        0           Coffee Shop                  Café   
29  Downtown Toronto        0           Coffee Shop                  Café   
28  Downtown Toronto        0           Coffee Shop                  Café   
19  Downtown Toronto        0           Coffee Shop                 Hotel   
11  Downtown Toronto        0           Coffee Shop    Italian Restaurant   
17  Downtown Toronto        0           Coffee Shop    Italian Restaurant   
12  Downtown Toronto        0           Coffee Shop   Japanese Restaurant   
13  Downtown Toronto        0           Coffee Shop                   Pub   
16  Downtown Toronto        0           Coffee Shop    Seafood Restaurant   
7    Central Toronto        0          Dessert Shop        Sandwich Place   
35      West Toronto        0             Gift Shop           Coffee Shop   
1       East Toronto        0      Greek Restaurant           Coffee Shop   
30  Downtown Toronto        0         Grocery Store                  Café   
5    Central Toronto        0                   Gym                 Hotel   
37      East Toronto        0  Gym / Fitness Center         Auto Workshop   
34      West Toronto        0    Mexican Restaurant       Thai Restaurant   
2       East Toronto        0                  Park            Board Shop   
0       East Toronto        0                  Park  Other Great Outdoors   
31      West Toronto        0              Pharmacy                Bakery   
9    Central Toronto        0                   Pub           Coffee Shop   
24   Central Toronto        0        Sandwich Place                  Café   

         3rd Most Common Venue      4th Most Common Venue  \
27            Airport Terminal                   Boutique   
32                  Restaurant                Men's Store   
4   Construction & Landscaping                Swim School   
26       Vietnamese Restaurant                Coffee Shop   
3          American Restaurant                     Bakery   
25                         Bar         Italian Restaurant   
14              Cosmetics Shop                       Café   
6                  Coffee Shop                Yoga Studio   
33                        Café      Performing Arts Venue   
18                         Bar                 Steakhouse   
21                       Hotel                 Restaurant   
20                       Hotel                 Restaurant   
15                       Hotel                 Restaurant   
36          Italian Restaurant           Sushi Restaurant   
29                  Restaurant                      Hotel   
28                  Restaurant         Seafood Restaurant   
19                    Aquarium                       Café   
11                        Café                        Pub   
17              Ice Cream Shop             Sandwich Place   
12            Sushi Restaurant                 Restaurant   
13                        Park      

## Cluster 1

In [51]:
toronto_central_clustered_df.loc[toronto_central_clustered_df['Cluster'] == 1, toronto_central_clustered_df.columns[[1] + list(range(5, toronto_central_clustered_df.shape[1]))]]

Borough  Cluster 1st Most Common Venue 2nd Most Common Venue  \
10  Downtown Toronto        1                  Park            Playground   
23   Central Toronto        1                  Park                 Trail   
8    Central Toronto        1            Playground                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
10                 Trail           Yoga Studio      Department Store   
23         Jewelry Store      Sushi Restaurant           Yoga Studio   
8            Yoga Studio      Department Store  Ethiopian Restaurant   

   6th Most Common Venue        7th Most Common Venue  \
10  Ethiopian Restaurant            Electronics Store   
23          Dessert Shop         Ethiopian Restaurant   
8      Electronics Store  Eastern European Restaurant   

          8th Most Common Venue        9th Most Common Venue  \
10  Eastern European Restaurant          Dumpling Restaurant   
23            Electronics Store  Eastern European Restaurant   
8           Dumpling Restaurant                   Donut Shop   

   10th Most Common Venue  
10             Donut Shop  
23    Dumpling Restaurant  
8        Doner Restaurant

## Cluster 2

In [52]:
toronto_central_clustered_df.loc[toronto_central_clustered_df['Cluster'] == 2, toronto_central_clustered_df.columns[[1] + list(range(5, toronto_central_clustered_df.shape[1]))]]

Borough  Cluster 1st Most Common Venue 2nd Most Common Venue  \
22  Central Toronto        2                Garden           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
22      Department Store           Event Space  Ethiopian Restaurant   

   6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
22     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   

   9th Most Common Venue 10th Most Common Venue  
22            Donut Shop       Doner Restaurant

<p>Upon observing the result, we can name the clusters as follow: </p>
<li>Cluster 0: Business area (with lots of business venues)</li>
<li>Cluster 1: Living area (with mostly park, trail, school; and some small businesses)</li>
<li>Cluster 2: Central Toronto (only a garden)</li>
